### PR TITLE
feat: 11 add shoppingcart

### PR DESCRIPTION
**Pull request by EduarfF**

### CODE REVIEW

**Review by bodya4243 — APPROVED** [review]
Well done

File: src/main/java/org/example/springbootintro/controller/ShoppingCartController.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    public CartItemDto addBookToShoppingCart(
+            @RequestBody @Valid CreateCartItemRequestDto addCartItemDto,
+            @AuthenticationPrincipal User user
+    ) {
+        return shoppingCartService.addCartItem(addCartItemDto, user);
+    }
+
+    @GetMapping
+    @Transactional
+    @PreAuthorize("hasRole('USER') or hasRole('ADMIN')")
```
Comment (karma-o) [inline src/main/java/org/example/springbootintro/controller/ShoppingCartController.java]: you can use `hasAnyRole()`

File: src/main/java/org/example/springbootintro/controller/ShoppingCartController.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+    @DeleteMapping("/cart-items/{cartItemsId}")
+    @PreAuthorize("hasRole('USER') or hasRole('ADMIN')")
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    @Operation(
+            summary = "Delete book from shopping cart",
+            description = "Delete book from shopping cart"
+    )
+    public void delete(
+            @PathVariable Long cartItemsId,
```
Comment (karma-o) [inline src/main/java/org/example/springbootintro/controller/ShoppingCartController.java]:
```suggestion
            @PathVariable Long cartItemId,
```

File: src/main/java/org/example/springbootintro/dto/cartitem/CreateCartItemRequestDto.java [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+import jakarta.validation.constraints.NotNull;
+import lombok.Data;
+
+@Data
+public class CreateCartItemRequestDto {
+    @NotNull(message = "Book ID can not be null")
+    private Long bookId;
+
+    @Min(value = 1, message = "Quantity must be greater than 0")
+    private int quantity;
```
Comment (karma-o) [inline src/main/java/org/example/springbootintro/dto/cartitem/CreateCartItemRequestDto.java:14]: what if user sends you null?

File: src/main/java/org/example/springbootintro/dto/cartitem/UpdateCartItemDto.java [hunk 4]
```diff
@@ -0,0 +1,10 @@
+package org.example.springbootintro.dto.cartitem;
+
+import jakarta.validation.constraints.Min;
+import lombok.Data;
+
+@Data
+public class UpdateCartItemDto {
+    @Min(value = 1, message = "Quantity must be greater than 0")
+    private int quantity;
```
Comment (karma-o) [inline src/main/java/org/example/springbootintro/dto/cartitem/UpdateCartItemDto.java:11]: same here

File: src/main/java/org/example/springbootintro/model/CartItem.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+import jakarta.persistence.Table;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.Setter;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.SQLRestriction;
+
+@Entity
+@Getter
+@Setter
```
Comment (karma-o) [inline src/main/java/org/example/springbootintro/model/CartItem.java:19]: do not forget about equals & hashcode & toString

File: src/main/java/org/example/springbootintro/model/CartItem.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+@Table(name = "cart_items")
+@SQLDelete(sql = "UPDATE cart_items SET is_deleted = true WHERE id =?")
+@SQLRestriction(value = "is_deleted=false")
```
Comment (karma-o) [inline src/main/java/org/example/springbootintro/model/CartItem.java]: wouldn't say that we need a soft delete for cart items, does not seem like a very important an irreplaceable information for me, and user can update their cart multiple times, remove some books from it several times a day, should we store all that? I'd say no

File: src/main/java/org/example/springbootintro/model/ShoppingCart.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+import lombok.Setter;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.SQLRestriction;
+
+@Entity
+@Getter
+@Setter
+@Table(name = "shopping_carts")
+@SQLDelete(sql = "UPDATE shopping_carts SET is_deleted = true WHERE id =?")
+@SQLRestriction(value = "is_deleted=false")
```
Comment (karma-o) [inline src/main/java/org/example/springbootintro/model/ShoppingCart.java:25]: I do not think we will ever delete a shopping cart, maybe if we want to delete a user, but I think that would be expected if we delete the user - the cart will be emptied and deleted as well, but that is more of a business question, so up to you

File: src/main/java/org/example/springbootintro/service/cartitem/CartItemServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+        ShoppingCart shoppingCart = getShoppingCartForUser(userId);
+        CartItem cartItem = getCartItem(cartItemId, shoppingCart.getId());
+        cartItemRepository.delete(cartItem);
+    }
+
+    @Override
+    public void deleteCartItemsAll(Long userId) {
+        ShoppingCart shoppingCart = getShoppingCartForUser(userId);
+        Set<CartItem> cartItem = cartItemRepository.findByShoppingCartId(shoppingCart.getId());
+        cartItemRepository.deleteAll(cartItem);
```
Comment (karma-o) [inline src/main/java/org/example/springbootintro/service/cartitem/CartItemServiceImpl.java]: here we do not need to check any cart item ids, so I think something like that would be better
```suggestion
        cartItemRepository.deleteAllByShoppingCartId(shoppingCart.getId());
```

File: src/main/java/org/example/springbootintro/service/cartitem/CartItemServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+                                "Can not find cart item by id:" + cartItemId)
+                );
+    }
+
+    private ShoppingCart getShoppingCartForUser(Long userId) {
+        return shoppingCartRepository.findByUserId(userId)
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Can not find shopping cart by id: " + userId)
+                );
+    }
```
Comment (karma-o) [inline src/main/java/org/example/springbootintro/service/cartitem/CartItemServiceImpl.java]: move that to shoppingCartService

File: src/main/java/org/example/springbootintro/service/shoppingcart/ShoppingCartServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+
+    @Override
+    public void deleteCartItemById(Long userId, Long cartItemId) {
+        cartItemService.deleteCartItemById(userId, cartItemId);
+    }
+
+    @Override
+    public void deleteCartItemsAll(Long userId) {
+        cartItemService.deleteCartItemsAll(userId);
+    }
```
Comment (karma-o) [inline src/main/java/org/example/springbootintro/service/shoppingcart/ShoppingCartServiceImpl.java]: we do not need to wrap 1 service methods into another one, we can just use them directly

**Review by karma-o — CHANGES_REQUESTED** [review]
We need to do some improvements to your solutin

**Review by karma-o — APPROVED** [review]
Nice!

**Pull request merged**